### PR TITLE
test(jobs): cover services.py functions

### PR DESCRIPTION
**Pull request by Negoje**

### CODE REVIEW

**Review by UkiStefanovic — APPROVED** [review]
Tests are working. Nice job!

File: annotation/tests/test_jobs.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            ANNOTATORS[1],
+        ],
+        validation_type=ValidationSchema.cross,
+        is_auto_distribution=False,
+        categories=CATEGORIES,
+        deadline="2021-10-19T01:01:01",
+        tenant=JOB_TENANT,
+    ),
+)
+# FILE_TEST_PROGRESS = File(
```
[the remark this answers is not in the document]
Comment (Negoje) [reply annotation/tests/test_jobs.py:73]: Yes, i put them there for future testing

File: annotation/.pre-commit-config.yaml [hunk 2]
```diff
@@ -19,7 +19,7 @@ repos:
     hooks:
       - id: flake8
         args:
-          - --extend-ignore=E203
+          - --extend-ignore=E203, F401
```
Comment (cakeinsauce) [inline annotation/.pre-commit-config.yaml:22]: Let's keep original settings

File: annotation/tests/test_jobs.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+#     pages_number=5,
+#     status=FileStatusEnumSchema.pending,
+# )
+# FILE_TEST_PROGRESS_3 = File(
+#     file_id=3,
+#     tenant=JOB_TENANT,
+#     job_id=JOB_IDS[2],
+#     pages_number=5,
+#     status=FileStatusEnumSchema.pending,
+# )
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:93]: Remove these comments, please

File: annotation/tests/test_jobs.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        ],
+        validation_type=ValidationSchema.cross,
+        is_auto_distribution=False,
+        categories=CATEGORIES,
+        deadline="2021-10-19T01:01:01",
+        tenant=JOB_TENANT,
+    ),
+    Job(  # Extraction job with no tasks
+        job_id=JOB_IDS[1],
+        callback_url="http://www.test.com",
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:51]: Could you make this testing data as a fixture?

File: annotation/tests/test_jobs.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+
+def test_update_job_status():
+
+    mock_session = MagicMock()
+
+    update_inner_job_status(
+        mock_session, JOB_IDS[0], JobStatusEnumSchema.finished
+    )
+
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:103]: remove empty lines, no need for them

File: annotation/tests/test_jobs.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    "validation_type," "annotators",
+    [
+        (ValidationSchema.cross, set([uuid.UUID(ANNOTATORS[0].user_id)])),
+        (ValidationSchema.hierarchical, set()),
+        (
+            ValidationSchema.validation_only,
+            set([uuid.UUID(ANNOTATORS[0].user_id)]),
+        ),
+    ],
+)
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:119]: Whenever you see collection which wont change during its lifetime consider using tuples instead of lists. And for parameter names use tuple please, instead of a single string - it is more readable

File: annotation/tests/test_jobs.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+@pytest.mark.unittest
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:120]: No need for this mark

File: annotation/tests/test_jobs.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+def test_check_annotators_cross(validation_type, annotators):
+    if validation_type == ValidationSchema.cross:
+        with pytest.raises(FieldConstraintError):
+            check_annotators(annotators, validation_type)
+    elif validation_type == ValidationSchema.hierarchical:
+        with pytest.raises(FieldConstraintError):
+            check_annotators(annotators, validation_type)
+    elif validation_type == ValidationSchema.validation_only:
+        with pytest.raises(FieldConstraintError):
+            check_annotators(annotators, validation_type)
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:130]: You basically have the same code in every if, what's the point of these clauses then?

File: annotation/tests/test_jobs.py (hunk 4 above)
Comment (Negoje) [reply annotation/tests/test_jobs.py]: I made the testing data as a fixture, but like this i can't use pytest.parametrize for "annotators" parameter in test_check_annotators.

File: annotation/tests/test_jobs.py (hunk 6 above)
Comment (Negoje) [reply annotation/tests/test_jobs.py]: Will do!

File: annotation/tests/test_jobs.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+
+@pytest.fixture
+def job_tenant():
+    return "test"
+
+
+@pytest.fixture
+def job_ids():
+    return (1, 2, 3)
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:24]: I would say it is too excessive, remove these fixtures, please

File: annotation/tests/test_jobs.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    User(user_id="ef81a4d0-cc01-447b-9025-a70ed441672d"),
+)
+
+
+@pytest.fixture
+def job_annotators():
+    return (
+        User(user_id="82533770-a99e-4873-8b23-6bbda86b59ae"),
+        User(user_id="ef81a4d0-cc01-447b-9025-a70ed441672d"),
+    )
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:38]: same data, remove constant please

File: annotation/tests/test_jobs.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+def categories():
+    return Category(
+        id="18d3d189e73a4680bfa77ba3fe6ebee5",
+        name="Test",
+        type=CategoryTypeSchema.box,
+    )
+
+
+@pytest.fixture
+def jobs_to_test_progress(annotators, categories):
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:51]: add type hints, please

File: annotation/tests/test_jobs.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+ANNOTATORS = (
+    User(user_id="82533770-a99e-4873-8b23-6bbda86b59ae"),
+    User(user_id="ef81a4d0-cc01-447b-9025-a70ed441672d"),
+)
+
+
+@pytest.fixture
+def job_annotators():
+    return (
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:35]: yield

File: annotation/tests/test_jobs.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+        User(user_id="82533770-a99e-4873-8b23-6bbda86b59ae"),
+        User(user_id="ef81a4d0-cc01-447b-9025-a70ed441672d"),
+    )
+
+
+@pytest.fixture
+def categories():
+    return Category(
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:43]: yield

Comment (cakeinsauce) [reply annotation/tests/test_jobs.py:43]: https://docs.pytest.org/en/6.2.x/fixture.html#yield-fixtures-recommended

File: annotation/tests/test_jobs.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+        id="18d3d189e73a4680bfa77ba3fe6ebee5",
+        name="Test",
+        type=CategoryTypeSchema.box,
+    )
+
+
+@pytest.fixture
+def jobs_to_test_progress(annotators, categories):
+    return (
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:52]: yield

File: annotation/tests/test_jobs.py [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+            status=FileStatusEnumSchema.pending,
+        ),
+        File(
+            file_id=3,
+            tenant=job_tenant,
+            job_id=jobs_to_test_progress[2].job_id,
+            pages_number=5,
+            status=FileStatusEnumSchema.pending,
+        ),
+    )
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:110]: do we need so much data for this fixture?  from a glance they are identical and don't have "edge case" scenarios. Same applies for jobs fixture

File: annotation/tests/test_jobs.py [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+
+def test_update_inner_job_status(job_ids):
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:113]: add type hints, please

File: annotation/tests/test_jobs.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+        "annotators",
+    ),
+    [
+        (ValidationSchema.cross, set((uuid.UUID(ANNOTATORS[0].user_id),))),
+        (ValidationSchema.hierarchical, set()),
+        (
+            ValidationSchema.validation_only,
+            set((uuid.UUID(ANNOTATORS[0].user_id),)),
+        ),
+    ],
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:135]: list -> tuple

Comment (cakeinsauce) [reply annotation/tests/test_jobs.py:135]: for sets you can use {} syntax when it is not empty

File: annotation/tests/test_jobs.py [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+)
+def test_check_annotators(validation_type, annotators):
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:137]: add type hints, please

File: annotation/tests/test_jobs.py (hunk 4 above)
Comment (cakeinsauce) [reply annotation/tests/test_jobs.py]: That's why you probably dont need a fixture and constant for this. create a simple user is not a big deal, `User(user_id="82533770-a99e-4873-8b23-6bbda86b59ae")`

File: annotation/tests/test_jobs.py (hunk 10 above)
Comment (Negoje) [reply annotation/tests/test_jobs.py:38]: I can't use in function "test_check_annotators" parametrize the fixture "job_annotators" that's why I have the constant one.

Comment (Negoje) [reply annotation/tests/test_jobs.py]: The reply to previous conversation didn't load for me. I just saw the reply there.

File: annotation/tests/test_jobs.py [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+    yield Category(
+        id="18d3d189e73a4680bfa77ba3fe6ebee5",
+        name="Test",
+        type=CategoryTypeSchema.box,
+    )
+
+
+@pytest.fixture
+def jobs_to_test_progress(
+    job_annotators: Tuple[User], categories: Tuple[Category]
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:41]: could you clarify why we yield tuples in those fixtures with a single value? Lets make it a single job, without tuple. Same applies for other fixtures. And here categories fixture is not Tuple, it is Category

File: annotation/tests/test_jobs.py [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+    ),
+    (
+        (
+            ValidationSchema.cross,
+            {uuid.UUID("82533770-a99e-4873-8b23-6bbda86b59ae")},
+        ),
+        (ValidationSchema.hierarchical, set()),
+        (
+            ValidationSchema.validation_only,
+            {uuid.UUID("82533770-a99e-4873-8b23-6bbda86b59ae")},
```
Comment (cakeinsauce) [inline annotation/tests/test_jobs.py:82]: users are the same, meaning we dont need them in parametrize

**Review by cakeinsauce — APPROVED** [review]
